### PR TITLE
Gi

### PR DESCRIPTION
**Pull request by e2002e**
Once again, a new branch (like my ganja plants), here will emission working.
Capture d’écran du 2022-11-28 17-17-56

### CODE REVIEW

File: Shaders/std/conetrace.glsl [hunk 1]
```diff
@@ -30,6 +30,100 @@ vec3 tangent(const vec3 n) {
...
+	col += traceCone(voxels, origin, mix(normal, c2, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -c2, angleMix), aperture, MAX_DISTANCE);
+	return (col / 9.0) * voxelgiOcc;
+	#endif
+
+	return vec4(0.0);
+}
+
+vec3 traceSpecular(sampler3D voxels, const vec3 pos, const vec3 normal, const vec3 viewDir, const float roughness) {
+	float specularAperture = clamp(tan((3.14159265 / 2) * roughness * 0.75), 0.0174533 * 3.0, 3.14159265);
```
Comment (LVutner) [inline Shaders/std/conetrace.glsl:114]: Tiny question. How did you come up with this formula? 
Does it approximate sort of microfacet brdf? 🤔

Comment (e2002e) [reply Shaders/std/conetrace.glsl:114]: hi, I did not change any of this. There are links referred in either make_voxels.py or this file;

File: blender/arm/make_renderpath.py [hunk 2]
```diff
@@ -283,7 +308,6 @@ def build():
             assets.add_shader_pass('blur_gaus_pass')
 
         if rpdat.rp_ssr:
-            wrd.world_defs += '_SSR'
```
Comment (MoritzBrueckner) [inline blender/arm/make_renderpath.py:286]: What's the deal with removing this line?

File: blender/arm/material/make_voxel.py [hunk 3]
```diff
@@ -1,29 +1,253 @@
...
-if arm.is_reload(__name__):
-    arm.utils = arm.reload_module(arm.utils)
-    assets = arm.reload_module(assets)
-    mat_state = arm.reload_module(mat_state)
-else:
-    arm.enable_reload(__name__)
-
+import arm.material.mat_utils as mat_utils
+import arm.material.make_particle as make_particle
+import arm.make_state as state
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_voxel.py:8]: Looks like the module reloading got lost here

File: Sources/armory/renderpath/Inc.hx [hunk 4]
```diff
@@ -538,23 +582,102 @@ class Inc {
...
-	#if arm_debug
-	public static var shadowsLogicTime = 0.0;
-	public static var shadowsRenderTime = 0.0;
-	static var startShadowsLogicTime = 0.0;
-	static var startShadowsRenderTime = 0.0;
-	static var callBackSetup = false;
-	static function setupEndFrameCallback() {
-		if (!callBackSetup) {
-			callBackSetup = true;
-			iron.App.endFrameCallbacks.push(endFrame);
```
Comment (MoritzBrueckner) [inline Sources/armory/renderpath/Inc.hx:550]: Is this supposed to be removed?

File: Sources/armory/renderpath/Inc.hx [hunk 5]
```diff
@@ -538,23 +582,102 @@ class Inc {
...
+	public static function computeVoxelsEnd() {
+	 	var rts = path.renderTargets;
+	 	var res = Inc.getVoxelRes();
+	 	path.generateMipmaps("voxels");
 	}
-	static function beginShadowsLogicProfile() { setupEndFrameCallback(); startShadowsLogicTime = kha.Scheduler.realTime(); }
-	static function beginShadowsRenderProfile() { startShadowsRenderTime = kha.Scheduler.realTime(); }
-	static function endShadowsLogicProfile() { shadowsLogicTime += kha.Scheduler.realTime() - startShadowsLogicTime - shadowsRenderTime; }
-	static function endShadowsRenderProfile() { shadowsRenderTime += kha.Scheduler.realTime() - startShadowsRenderTime; }
-	public static function endFrame() { shadowsLogicTime = 0;  shadowsRenderTime = 0; }
```
Comment (MoritzBrueckner) [inline Sources/armory/renderpath/Inc.hx:557]: Same here

File: blender/arm/make_renderpath.py [hunk 6]
```diff
@@ -301,19 +325,31 @@ def build():
         assets.add_khafile_def('rp_stereo')
         assets.add_khafile_def('arm_vr')
         wrd.world_defs += '_VR'
+        assets.add(assets_path + 'vr.png')
+        assets.add_embedded_data('vr.png')
```
Comment (MoritzBrueckner) [inline blender/arm/make_renderpath.py:329]: Is this related to voxel GI?

File: blender/arm/make_renderpath.py (hunk 2 above)
Comment (e2002e) [reply blender/arm/make_renderpath.py:286]: I added that myself and it got included in one of the pull requests.
I'm pretty sure, verified on older version and this line's not present.

File: blender/arm/material/make_voxel.py (hunk 3 above)
Comment (e2002e) [reply blender/arm/material/make_voxel.py:8]: Ok

File: blender/arm/make_renderpath.py (hunk 6 above)
Comment (e2002e) [reply blender/arm/make_renderpath.py:329]: I have no clue. Anyway this file doesn't exist in current version.

File: Shaders/std/conetrace.glsl (hunk 1 above)
Comment (LVutner) [reply Shaders/std/conetrace.glsl:114]: Hm, seems like we might be able to approximate GGX, to match reflections & BRDF used in Armory.
Thanks for mentioning those links!

**Comment by e2002e** [comment]
Hi @luboslenco, just reminding you that you merged the iron part of the code. You should get this one too.

File: blender/arm/material/make_voxel.py [hunk 7]
```diff
@@ -11,19 +14,247 @@
...
+        frag.write('basecol = vec3(0.0);')
+
+    frag.write('vec3 voxel = voxposition * 0.5 + 0.5;')
+    frag.write('imageStore(voxels, ivec3(voxelgiResolution * voxel), vec4(min(basecol, vec3(1.0)), 1.0));')
+
+    return con_voxel
+
 def make_ao(context_id):
-    con_voxel = mat_state.data.add_context({ 'name': context_id, 'depth_write': False, 'compare_mode': 'always', 'cull_mode': 'none', 'color_writes_red': [False], 'color_writes_green': [False], 'color_writes_blue': [False], 'color_writes_alpha': [False], 'conservative_raster': False })
+    con_voxel = mat_state.data.add_context({ 'name': context_id, 'depth_write': False, 'compare_mode': 'always', 'cull_mode': 'none', 'color_write_red': False, 'color_write_green': False, 'color_write_blue': False, 'color_write_alpha': False, 'conservative_raster': False })
```
Comment (luboslenco) [inline blender/arm/material/make_voxel.py:257]: Why this change? The `color_writes_red` should be the correct one.

File: Shaders/deferred_light/deferred_light.frag.glsl [hunk 8]
```diff
@@ -224,39 +230,35 @@ void main() {
...
 	#ifdef _Rad
 		prefilteredColor = pow(prefilteredColor, vec3(2.2));
 	#endif
 #endif
 
+#ifdef _Rad
+	vec3 reflectionWorld = reflect(-v, n);
+	float lod = getMipFromRoughness(roughness, envmapNumMipmaps);
+	vec3 prefilteredColor = textureLod(senvmapRadiance, envMapEquirect(reflectionWorld), lod).rgb;
+#endif
```
Comment (luboslenco) [inline Shaders/deferred_light/deferred_light.frag.glsl:261]: Hm why still move the block? Just from a glance there is `vec3 prefilteredColor`, but `prefilteredColor` is already used above?

**Comment by e2002e** [comment]
Hi, I corrected what was discussed after review.
You can merge that if it seems right.

File: blender/arm/material/make_voxel.py [hunk 9]
```diff
@@ -11,19 +14,247 @@
...
+        con = make_ao(context_id)
 
     assets.vs_equal(con, assets.shader_cons['voxel_vert'])
     assets.fs_equal(con, assets.shader_cons['voxel_frag'])
     assets.gs_equal(con, assets.shader_cons['voxel_geom'])
 
     return con
 
+def make_gi(context_id):
+    con_voxel = mat_state.data.add_context({ 'name': context_id, 'depth_write': False, 'compare_mode': 'always', 'cull_mode': 'none', 'color_write_reds': [False], 'color_writes_green': [False], 'color_writes_blue': [False], 'color_write_alpha': False, 'conservative_raster': True })
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_voxel.py:31]: There's a small typo here: `color_write_reds` instead of `color_writes_red` I assume.

File: Shaders/deferred_light/deferred_light.frag.glsl [hunk 10]
```diff
@@ -513,6 +535,17 @@ void main() {
...
+	/*
+	vec3 origin = vec3(texCoord * 2.0 - 1.0, 0.99);
+	vec3 direction = vec3(0.0, 0.0, -1.0);
+	vec4 color = vec4(0.0f);
+	for(uint step = 0; step < 400 && color.a < 0.99f; ++step) {
+	 	vec3 point = origin + 0.005 * step * direction;
+	 	color += (1.0f - color.a) * textureLod(voxels, point * 0.5 + 0.5, 0);
+	}
+	fragColor.rgb += color.rgb;
+	*/
```
Comment (MoritzBrueckner) [inline Shaders/deferred_light/deferred_light.frag.glsl:550]: Is this for debugging (👍 )? If so, what do you think of adding a `#define DEBUG_VOXELS 1` at the top of the shader so that it can be easily toggled?

**Comment by MoritzBrueckner** [comment]
As I mentioned in one of your earlier PRs I can't judge the implementation since I don't know much about voxel GI, but apart from the two very minor comments above the code looks good to me now :)

File: Shaders/deferred_light/deferred_light.frag.glsl (hunk 10 above)
Comment (e2002e) [reply Shaders/deferred_light/deferred_light.frag.glsl:550]: Yes that is for debug, no It's better leave it like this. Lubos had used that before when creating the shaders I guess.

**Comment by e2002e** [comment]
Thanks for the review, I'm leaving the pull request as is now.

File: Shaders/deferred_light/deferred_light.frag.glsl [hunk 11]
```diff
@@ -373,12 +396,6 @@ void main() {
 		#endif
 	#endif
 
-	#ifdef _VoxelAOvar
-	#ifdef _VoxelShadow
-	svisibility *= 1.0 - traceShadow(voxels, voxpos, sunDir);
```
Comment (luboslenco) [inline Shaders/deferred_light/deferred_light.frag.glsl:378]: Why remove this? Is `_VoxelShadow` still working?

Comment (e2002e) [reply Shaders/deferred_light/deferred_light.frag.glsl:378]: I don't know I can't see a difference when activated or not.
I have set these functions to use this feature even with global illumination but that was just a bet.
No impairing anyway.

**Comment by e2002e** [comment]
screen

**Comment by e2002e** [comment]
@luboslenco I'm reminding you that you merge iron that will not run without this.
This is working fine. You can just merge it if you wish.

File: blender/arm/make_renderpath.py [hunk 12]
```diff
@@ -91,11 +105,19 @@ def add_world_defs():
...
-        wrd.world_defs += '_VoxelAOvar' # Write a shader variant
-        if rpdat.arm_voxelgi_shadows:
-            wrd.world_defs += '_VoxelShadow'
-        if rpdat.arm_voxelgi_occ == 0.0:
-            wrd.world_defs += '_VoxelAONoTrace'
+
+        if voxelgi:
+            wrd.world_defs += '_VoxelGI'
+            if rpdat.arm_voxelgi_shadows:
+                wrd.world_defs += '_VoxelGIShadow'
```
Comment (luboslenco) [inline blender/arm/make_renderpath.py:112]: Do we need to have a separate `_VoxelGIShadow` define? Looks like it's the same as `_VoxelShadow`.

File: Sources/armory/renderpath/Inc.hx [hunk 13]
```diff
@@ -444,15 +444,19 @@ class Inc {
 
 	#if rp_voxels
 	public static function initGI(tname = "voxels") {
-		#if arm_config
-		var config = armory.data.Config.raw;
-		if (config.rp_gi != true || voxelsCreated) return;
-		voxelsCreated = true;
-		#end
-
```
Comment (luboslenco) [inline Sources/armory/renderpath/Inc.hx:452]: Why remove this code?

File: blender/arm/make_renderpath.py (hunk 12 above)
Comment (e2002e) [reply blender/arm/make_renderpath.py:112]: This is right, just that the old files did differentiate that

**Comment by luboslenco** [comment]
> I'm reminding you that you merge iron that will not run without this.

I tested with the latest armsdk and voxelao is still working, iirc we just changed the name of the voxel define.

File: blender/arm/make_renderpath.py (hunk 12 above)
Comment (luboslenco) [reply blender/arm/make_renderpath.py]: What old files? Which revision should I look at? Still, if these defines are the same we should merge them into one.

Comment (e2002e) [reply blender/arm/make_renderpath.py]: I think the voxels gi comes from 19.02

File: Sources/armory/renderpath/Inc.hx [hunk 14]
```diff
@@ -531,12 +545,20 @@ class Inc {
 	}
 
 	public static inline function getVoxelResZ(): Float {
-		#if (rp_voxelgi_resolution_z == 1.0)
```
Comment (luboslenco) [inline Sources/armory/renderpath/Inc.hx:534]: If we need this many, maybe just output the float value for `rp_voxelgi_resolution_z` directly alongside these lines? https://github.com/armory3d/armory/blob/main/blender/arm/write_data.py#L468

File: blender/arm/make_renderpath.py (hunk 2 above)
Comment (luboslenco) [reply blender/arm/make_renderpath.py:295]: > I added that myself and it got included in one of the pull requests

Hm, can you please point to the pull request where it got included?

File: blender/arm/make_renderpath.py [hunk 15]
```diff
@@ -27,9 +27,11 @@ def add_world_defs():
     # Screen-space ray-traced shadows
     if rpdat.arm_ssrs:
         wrd.world_defs += '_SSRS'
+        assets.add_khafile_def('rp_ssrs')
 
     if rpdat.arm_micro_shadowing:
         wrd.world_defs += '_MicroShadowing'
+        assets.add_khafile_def('rp_microshadowing')
```
Comment (luboslenco) [inline blender/arm/make_renderpath.py:34]: Why is this added?

File: blender/arm/make_renderpath.py [hunk 16]
```diff
@@ -129,8 +151,11 @@ def add_world_defs():
         wrd.world_defs += '_Clusters'
         assets.add_khafile_def('arm_clusters')
 
-    if '_Rad' in wrd.world_defs and '_Brdf' not in wrd.world_defs:
+	#really unsure about these lines.
```
[the remark this answers is not in the document]
Comment (e2002e) [reply blender/arm/make_renderpath.py:154]: richtig, indPos is unused.

File: blender/arm/make_renderpath.py (hunk 15 above)
Comment (e2002e) [reply blender/arm/make_renderpath.py:34]: i really do not remember, might had something to do with one of my first pull request fixing translucency and/or tesselation with microshadowing.

File: Sources/armory/renderpath/Inc.hx (hunk 14 above)
Comment (luboslenco) [reply Sources/armory/renderpath/Inc.hx:534]: > There still could be a change of resolution while playing

Can you elaborate on that? It looks like this function just converts the compile-time `rp_voxelgi_resolution_z` define into float.

Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:534]: I was thinking of the live patch mode.
Res is written to compiled.inc, read from renderpath and converted to floats.
writing float directly into the file will not change the need for a function in Inc.

Comment (luboslenco) [reply Sources/armory/renderpath/Inc.hx:534]: The `#if`s are resolved at compile time, nothing to do with performance. I was asking to replace multiple source code lines with a single one. 🙂

Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:534]: It's strange that we can test rp_voxelgi_resolution against an integer via a macro, but can't return it's value straight away.

This would go like:```

public static inline function getVoxelRes(): Int {
	return rp_voxelgi_resolution;
}
```

`Unknown identifier : rp_voxelgi_resolution`

Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:534]: I will look at that at the end of the afternoon

File: blender/arm/make_renderpath.py (hunk 2 above)
Comment (e2002e) [reply blender/arm/make_renderpath.py]: It's from the water pass changes, since we need to know whether we'r using ssr or not.
Fixed it !

File: Sources/armory/renderpath/Inc.hx (hunk 14 above)
Comment (e2002e) [reply Sources/armory/renderpath/Inc.hx:534]: I'm not able to get preprocessor expressions values.
This change is irrelevant though, if you need to fix this do so.

**Comment by e2002e** [comment]
Thanks for the review. I'm leaving the unearthed little girl into you hands. Aufviedersehen.

**Comment by e2002e** [comment]
Ich wunsche inhen eine frauh neue yahre.

**Comment by e2002e** [comment]
> > I'm reminding you that you merge iron that will not run without this.
> 
> I tested with the latest armsdk and voxelao is still working, iirc we just changed the name of the voxel define.

You merge this when you want.

**Review by e2002e — COMMENTED** [review]
this won't work and produce redish light

**Comment by MoritzBrueckner** [comment]
Hi, did you merge the changes of your now closed refraction PR into the branch of this PR? Looks like the refraction code is now part of this, but it doesn't really belong here since this is about Voxel GI...

**Comment by e2002e** [comment]
Yep, got included. Probably while I was trying to merge all changes to my 'main".
I do not know how to deal with that as of now. I'm just going to try and make the forward render work for refraction and do a commit with both features next time.
